### DISCUSSION TITLE
Modeling received messages as `AsyncStream`

### DISCUSSION THREAD

**Discussion by nikitamounier** [insidegui/MultipeerKit#27] [General]
Hi Rambo, 

First off, thank you for making this library – MPC is hard work, but this makes it so much easier. I'm trying to phase my code towards using Swift concurrency, and was wondering whether this is correct and doesn't disrespect any assumptions made in the library:

```swift
extension MultipeerTransceiver {
    func receive<T: Codable>(_ type: T.Type) -> AsyncStream<(T, Peer)> {
        return AsyncStream { continuation in
            self.receive(T.self) { payload, sender in
                continuation.yield((payload, sender))
            }
        }
    }
}

```
I saw that you have a draft PR for supporting Swift Concurrency, but I also saw that you hadn't made a "Swift Concurrency" version of `MultipeerTransceiver/receive(_:_:)`, so maybe there's something here that I'm missing.

This would enable something like this, where here I'm juggling with both MultipeerKit and [NearbyInteraction](https://developer.apple.com/documentation/nearbyinteraction):

```swift
for await (payload, sender) in transceiver.receive(Message.self) {
    
    switch payload {
    case let .discoveryToken(data):
        let token = try! NSKeyedUnarchiver.unarchivedObject(ofClass: NIDiscoveryToken.self, from: data)!
        
        let newSession = NISession()
        
        let sessionEvents = AsyncStream<Delegate.Event> { continuation in
            let delegate: Delegate? = Delegate(continuation)
            newSession.delegate = delegate
        }
        
        newSession.run(NINearbyPeerConfiguration(peerToken: token))
        peerSessions[sender.id] = newSession
        
        for await event in sessionEvents {
            switch event {
            // ...
            }
        }
              
    case let .profile(profile):
        self.peerProfile = profile
    }
}
```

Which I find super clean!

Let me know what you think.

**Comment by insidegui** [comment]
That PR is WIP, hence why it's missing API.

I don't have the time to analyze this in depth right now, but it looks good from what I've seen in your example. This is definitely something that I want to have in the official API soon, since I'm also using Swift concurrency in my own projects.

If you feel like opening a PR, perhaps picking up from the existing one, feel free to do so!
